synapse/federation/federation_client.py

```diff
@@ -1152,23 +1153,30 @@ async def _execute(pdu: EventBase) -> None:
                     % (auth_chain_create_events,)
                 )
 
-            if response.members_omitted and not response.servers_in_room:
-                raise InvalidResponseError(
-                    "members_omitted was set, but no servers were listed in the room"
-                )
+            servers_in_room = response.servers_in_room
+            if response.members_omitted:
+                if not servers_in_room:
+                    raise InvalidResponseError(
+                        "members_omitted was set, but no servers were listed in the room"
+                    )
 
-            if response.members_omitted and not partial_state:
-                raise InvalidResponseError(
-                    "members_omitted was set, but we asked for full state"
-                )
+                if destination not in servers_in_room:
+                    # `servers_in_room` is supposed to be a complete list.
+                    # Fix things up if the remote homeserver is badly behaved.
+                    servers_in_room = [destination] + servers_in_room
+
+                if not partial_state:
+                    raise InvalidResponseError(
+                        "members_omitted was set, but we asked for full state"
+                    )
```

Comment:
I'd recommend moving this above the destination check to have all the error checking up-front.

Reply:
Done in 163c68c0be.